Added stratify option to train_test_split function.

nandwalritik opened the pull request:
This PR adds `stratify` option to `train_test_split` method. I took reference from scikit-learn's `StratifiedShuffleSplit` class for implementing stratified split and integrated the changes as were suggested by @lhoestq.

It fixes #3452.

@lhoestq Please review and let me know, if any changes are required.

lhoestq commented on src/datasets/arrow_dataset.py:
```diff
@@ -3436,15 +3438,22 @@ def train_test_split(
+            # stratified partition
+            if stratify is not None:
+                train_indices, test_indices = next(
+                    generate_indices(self[stratify], n_train, n_test, random_state=seed)
```
This will boost performance a bit ;)
Arrow does zero-copy conversion to numpy

```suggestion
                    generate_indices(self.with_format("numpy")[stratify], n_train, n_test, random_state=seed)
```

lhoestq replied:
We instantiate the random number generator above using the seed already
```suggestion
                    generate_indices(self[stratify], n_train, n_test, random_state=generator)
```

nandwalritik replied:
changed as per the suggestions.

lhoestq commented on src/datasets/splits.py:
```diff
@@ -569,3 +572,135 @@ def __post_init__(self):
+    return floored.astype(np.int)
+
+
+def generate_indices(y, n_train, n_test, random_state=None, n_splits=10):
```
I think you can have this function in `arrow_dataset.py` directly, but use a more explicit and private name like `_stratified_shuffle_split_generate_indices`.

You can also move the other functions you added in `arrow_dataset.py`

nandwalritik replied:
Moved all the functions inside `arrow_dataset.py` and changed name of `generate_indices` to `_stratified_shuffle_split_generate_indices`.

lhoestq commented on src/datasets/splits.py:
```diff
@@ -569,3 +572,135 @@ def __post_init__(self):
+        np.sum(sampled_classes) == n_draws
+
+    """
+    rng = check_random_state(rng)
```
I think you can remove this and `check_random_state` since we always pass a random number generator from `train_test_split`

nandwalritik replied:
removed

lhoestq commented on src/datasets/splits.py:
```diff
@@ -569,3 +572,135 @@ def __post_init__(self):
+            "The test_size = %d should be greater or " "equal to the number of classes = %d" % (n_test, n_classes)
+        )
+    class_indices = np.split(np.argsort(y_indices, kind="mergesort"), np.cumsum(class_counts)[:-1])
+    rng = check_random_state(random_state)
```
same here

nandwalritik replied:
Fixed it.

lhoestq reviewed (commented):
Nice thank you ! This will be super useful :)

Could you also add some tests in test_arrow_dataset.py and add an example of usage in the `Example:` section of the `train_test_split` docstring ?

nandwalritik commented:
> Nice thank you ! This will be super useful :)
> 
> Could you also add some tests in test_arrow_dataset.py and add an example of usage in the `Example:` section of the `train_test_split` docstring ?

I will try to do it, is there any documentation for adding test cases? I have never done it before.

lhoestq commented:
Thanks for the changes !

> I will try to do it, is there any documentation for adding test cases? I have never done it before.

You can just add a function `test_train_test_split_startify` in `test_arrow_dataset.py`.

In this function you can define a dataset and make sure that `train_test_split` with the `stratify` argument works as expected.

You can do `pytest tests/test_arrow_dataset.py::test_train_test_split_startify` to run your test.

Feel free to get some inspiration from other tests like `test_interleave_datasets` for example

nandwalritik commented:
I have added tests for stratified train_test_split in `test_arrow_dataset.py` file inside `test_train_test_split_startify` function. I have also added example usage with `stratify` arg in `Example:` section of the `train_test_split` docstring.
Results of tests:
```
(data) nandwalritik@hp:~/datasets$ pytest tests/test_arrow_dataset.py::test_train_test_split_startify -W ignore
============================================================================ test session starts ============================================================================
platform linux -- Python 3.9.5, pytest-7.1.2, pluggy-1.0.0
rootdir: /home/nandwalritik/datasets
plugins: datadir-1.3.1, forked-1.4.0, xdist-2.5.0
collected 1 item                                                                                                                                                            

tests/test_arrow_dataset.py .                                                                                                                                         [100%]

============================================================================= 1 passed in 0.12s =============================================================================

```

mariosasko commented on src/datasets/arrow_dataset.py:
```diff
@@ -3280,6 +3391,7 @@ def train_test_split(
                 If int, represents the absolute number of train samples.
                 If None, the value is automatically set to the complement of the test size.
             shuffle (:obj:`bool`, optional, default `True`): Whether or not to shuffle the data before splitting.
+            stratify (:obj:`str`, optional, default `None`): data is split in a stratified fashion, using the specified class labels
```
Better docstring:
```suggestion
            stratify (:obj:`str`, optional, default `None`): The column name of labels to use to stratify data.
```

Let's add the `isinstance(stratify, datasets.ClassLabel)` check to the function body

mariosasko replied:
Oops, you are right. I mean this: `isinstance(self.features[stratify], ClassLabel)`

nandwalritik replied:
Yeah

mariosasko commented on src/datasets/arrow_dataset.py:
```diff
@@ -3319,6 +3431,24 @@ def train_test_split(
+            features: ['text', 'label'],
+            num_rows: 25000
+        })
+        >>> ds = ds.train_test_split(test_size=0.2,stratify="label")
```
```suggestion
        >>> ds = ds.train_test_split(test_size=0.2, stratify="label")
```

mariosasko commented on src/datasets/arrow_dataset.py:
```diff
@@ -3242,6 +3242,116 @@ def shuffle(
+        n_classes = classes.shape[0]
+        class_counts = np.bincount(y_indices)
+        if np.min(class_counts) < 2:
+            raise ValueError(
+                "The least populated class in y has only 1"
+                " member, which is too few. The minimum"
+                " number of groups for any class cannot"
+                " be less than 2."
```
Let's improve this error message by specifying the actual `stratify` column name instead of `y`

nandwalritik replied:
Fixed

mariosasko reviewed (commented):
Very nice!

PS: Maybe we should put the `_approximate_mode` and `_stratified_shuffle_split_generate_indices` in a separate module @lhoestq (`utils/stratify.py` for instance)? And if we keep them here as methods, we should at least mark them with the `@staticmethod` decorator.

lhoestq commented:
Thanks a lot !

`utils/stratify.py` sounds good yes :)

nandwalritik commented:
Added all the changes as were suggested and rebased with `main`.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/datasets/pr_4322). All of your documentation changes will be reflected on that endpoint.

nandwalritik commented on src/datasets/utils/stratify.py:
```diff
@@ -0,0 +1,112 @@
+    n_classes = classes.shape[0]
+    class_counts = np.bincount(y_indices)
+    if np.min(class_counts) < 2:
+        raise ValueError(
+            f"The least populated class in {stratify} column has only 1"
+            " member, which is too few. The minimum"
+            " number of groups for any class cannot"
+            " be less than 2."
+        )
```
It's clear to me, I will try adding as recommended.

nandwalritik replied:
I have made changes as you suggested, let me know if any more changes are required.

nandwalritik commented on tests/test_arrow_dataset.py:
```diff
@@ -3553,3 +3554,32 @@ def test_task_text_classification_when_columns_removed(self):
                 self.assertDictEqual(dset.features, features_after_map)
+
+
+def test_train_test_split_startify():
```
Added tests for testing error messages inside `test_errors_train_test_split_stratify` function of `StratifiedTest` class. I also moved `test_train_test_split_startify` function inside same class so that all the functions testing stratified split comes under same class.

mariosasko reviewed (approved):
Thanks! LGTM!

Pinging @lhoestq for the final review.

lhoestq commented on src/datasets/arrow_dataset.py:
```diff
@@ -3281,6 +3283,7 @@ def train_test_split(
                 If int, represents the absolute number of train samples.
                 If None, the value is automatically set to the complement of the test size.
             shuffle (:obj:`bool`, optional, default `True`): Whether or not to shuffle the data before splitting.
+            stratify_by_column (:obj:`str`, optional, default `None`):The column name of labels to be used to perform stratified split of data.
```
```suggestion
            stratify_by_column (:obj:`str`, optional, default `None`): The column name of labels to be used to perform stratified split of data.
```

lhoestq reviewed (approved):
Thanks a lot !

Damon03 commented:
Hi, I encounter an error when I try to specify the stratify_by_column. However, I have a columns which specific the label of the row as a string. But an error showed when I try to do it. "ValueError: Stratifying by column is only supported for ClassLabel column, and column code is Value.".

lhoestq commented:
Hi @Damon03 , you can change the type of your column to ClassLabel using
```python
ds = ds.class_encode_column(column_name)
```
then you'll be free to use `stratify` :)

Damon03 commented:
Thank you so much. It worked.
